File: src/server/web_request.cpp
/*
 WebServe
 Copyright 2018-2019 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#include "web_request.h"

#include <sstream>
#include <iostream>
#include <vector>

#include <cstdio>
#include <cstring>
#include <cstdlib>

#include "utils/string_helpers.h"

WebRequest::WebRequest(const std::string& rawRequest) : m_rawRequest(rawRequest),
	m_requestType(eRequestUnknown),
    m_httpVersion(eHTTPUnknown),
	m_connectionType(eConnectionUnknown),
	m_fileType(eFTUnknown),
	m_headerAuthenticationType(eAuthNone)
{

}

bool WebRequest::parse(Logger& logger)
{
	// TODO: doing it this way is very hacky as it means we lose blank lines (end of header, etc)...
	// TODO: also, allocating loads of separate strings for each line is fairly expensive - use some stringView type of thing?
	std::vector<std::string> lines;
	StringHelpers::split(m_rawRequest, lines);

	if (lines.empty())
		return false;

	// first line should contain main request
	const std::string &line = lines[0];

	size_t httpVersionItemStart = line.find("HTTP/");
	if (httpVersionItemStart == std::string::npos)
		return false;

	std::string httpVersionString = line.substr(httpVersionItemStart + 5, 3);
	if (httpVersionString == "1.1")
	{
		m_httpVersion = eHTTP11;
	}
	else if (httpVersionString == "1.0")
	{
		m_httpVersion = eHTTP10;
	}
	else if (httpVersionString == "0.9")
	{
		m_httpVersion = eHTTP09;
	}

	if (!StringHelpers::beginsWithStaticConst(line, "GET"))
	{
		if (!StringHelpers::beginsWithStaticConst(line, "POST"))
		{
			if (!StringHelpers::beginsWithStaticConst(line, "HEAD"))
			{
				std::string command;
				size_t commandSep = line.find(' ');
				if (commandSep != std::string::npos)
				{
					command = line.substr(0, commandSep);
				}
				
				m_requestType = eRequestUnknown;
	
				logger.error("Unsupported HTTP command in request from client: %s", command.c_str());
			}
			else
			{
				m_requestType = eRequestHEAD;
				
				// we recognise this, but don't handle it (for the moment)
			}
			
			return false;
		}

		m_requestType = eRequestPOST;
	}
	else
	{
		// it was a
		m_requestType = eRequestGET;		
	}

	size_t pathStart = m_requestType != eRequestGET ? 5 : 4;

	size_t pathEnd = line.rfind(' '); // should be the space before the HTTP version

	if (pathEnd == std::string::npos)
		return false;

	const std::string path = line.substr(pathStart, pathEnd - pathStart);

	size_t nQuestionMark = path.find('?');
	if (nQuestionMark == std::string::npos)
	{
		m_path = path;
	}
	else
	{
		m_path = path.substr(0, nQuestionMark);

		// if POST has this on the end, ignore the URL params
		if (m_requestType != eRequestPOST)
		{
			std::string params = path.substr(nQuestionMark + 1);

			addParams(params);
		}
	}

	// try and work out the file type
	if (m_requestType != eRequestPOST)
	{
		size_t extensionSep = m_path.find_last_of('.');
		if (extensionSep != std::string::npos)
		{
			std::string extensionString = m_path.substr(extensionSep + 1);
			if (extensionString == "css")
			{
				m_fileType = eFTCSS;
			}
		}
	}

	std::vector<std::string>::const_iterator itLine = lines.begin();
	++ itLine; // skip the first line which we've processed already


	bool lookForAuthentication = false;
	bool foundCookie = false;
	bool foundUserAgent = false;
	bool foundHost = false;
	bool foundConnection = false;

	// TODO: this needs to cope with case-insensitive comparisons...
	//       We can't just make the entire line lower-case though, it needs to be itemised

	for (; itLine != lines.end(); ++itLine)
	{
		const std::string& otherLine = *itLine;

		// now try and find an authentication line...
		if (lookForAuthentication && (otherLine.compare(0, 14, "Authorization:") == 0))
		{
			std::string authorizationString = otherLine.substr(15);

			processAuthenticationHeader(authorizationString);
			lookForAuthentication = false;
		}
		else if (!foundCookie && (otherLine.compare(0, 7, "Cookie:") == 0))
		{
			std::string cookieString;
			cookieString = extractFieldItem(otherLine, 7 + 1);

			processCookieHeader(cookieString);
			foundCookie = true;
		}
		else if (!foundUserAgent && (otherLine.compare(0, 11, "User-Agent:") == 0))
		{
			m_userAgentField = extractFieldItem(otherLine, 11 + 1);

			foundUserAgent = true;
		}
		else if (!foundHost && (otherLine.compare(0, 5, "Host:") == 0))
		{
			m_hostValue = extractFieldItem(otherLine, 5 + 1);

			foundHost = true;
		}
		else if (!foundConnection && (otherLine.compare(0, 11, "Connection:") == 0))
		{
			m_connectionValue = extractFieldItem(otherLine, 11 + 1);

			foundConnection = true;
		}
	}

	// apply defaults
	if (m_httpVersion == eHTTP11)
	{
		m_connectionType = eConnectionKeepAlive;
	}
	else if (m_httpVersion == eHTTP09 || m_httpVersion == eHTTP10)
	{
		m_connectionType = eConnectionClose;
	}

	if (!m_connectionValue.empty() && (m_httpVersion == eHTTP10 || m_httpVersion == eHTTP11))
	{
		std::string lowerCaseConnectionValue = StringHelpers::toLower(m_connectionValue);
		if (lowerCaseConnectionValue == "close")
		{
			m_connectionType = eConnectionClose;
		}
		else if (lowerCaseConnectionValue == "keep-alive")
		{
			m_connectionType = eConnectionKeepAlive;
		}
		else
		{
			logger.warning("Unknown connection type specified: %s", m_connectionValue.c_str());
		}
	}

	// TODO: this is crap - we should check for the double CR and use that to detect the end of the header...
	//       or as a hacky stop-gap to be more robust (and detect missing POST data), match the line
	//       by the Content-Length....
	if (m_requestType == eRequestPOST && lines.size() > 1)
	{
		// hopefully, last line of HTTP request should be the POST params

		const std::string& params = lines.back();

		// Note: Sometimes we don't seem to have any POST data, as WebKit always seems to send POST webform
		//       data after the data start in a second packet and our single receive doesn't cope with this...

		addParams(params);
	}

	return true;
}

bool WebRequest::hasParam(const std::string &name) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aParams.find(name);

	return itFind != m_aParams.end();
}

std::string WebRequest::getParam(const std::string& name) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aParams.find(name);
	
	if (itFind == m_aParams.end())
		return "";

	return itFind->second;
}

int WebRequest::getParamAsInt(const std::string& name, int defaultVal) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aParams.find(name);
	
	if (itFind == m_aParams.end())
		return defaultVal;
	
	std::string paramValue = itFind->second;

	if (paramValue.empty())
	{
		return defaultVal;
	}

	return atoi(paramValue.c_str());
}

// Note: this rebuilds the params in alphabetical order, instead of the original
//       order they were in from the original request.
std::string WebRequest::getParamsAsGETString(bool ignorePaginationParams) const
{
	std::string paramString;
	
	for (const auto& itParam : m_aParams)
	{
		const std::string& paramName = itParam.first;
		const std::string& paramValue = itParam.second;
		
		if (ignorePaginationParams)
		{
			if (paramName == "perPage" || paramName == "startIndex")
				continue;
		}
		
		if (!paramString.empty())
			paramString += "&";
		
		paramString += paramName + "=";
		paramString += StringHelpers::simpleEncodeString(paramValue);
	}
	
	return paramString;
}

bool WebRequest::hasCookie(const std::string& name) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aCookies.find(name);

	return itFind != m_aCookies.end();
}

std::string WebRequest::getCookie(const std::string& name) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aCookies.find(name);

	if (itFind == m_aCookies.end())
		return "";

	return itFind->second;
}

int WebRequest::getCookieAsInt(const std::string& name, int defaultVal) const
{
	std::string cookieValue = getCookie(name);

	if (cookieValue.empty())
	{
		return defaultVal;
	}

	return atoi(cookieValue.c_str());
}

int WebRequest::getParamOrCookieAsInt(const std::string& paramName, const std::string& cookieName, int defaultValue) const
{
	std::map<std::string, std::string>::const_iterator itFind = m_aParams.find(paramName);

	if (itFind != m_aParams.end())
	{
		std::string paramValue = itFind->second;
		if (!paramValue.empty())
		{
			return atoi(paramValue.c_str());
		}
	}

	// otherwise, look for the cookie...
	std::string cookieValue = getCookie(cookieName);

	if (cookieValue.empty())
	{
		return defaultValue;
	}

	return atoi(cookieValue.c_str());
}

std::string WebRequest::extractFieldItem(const std::string& fieldString, unsigned int valueStartPos)
{
	std::string fieldValue;
	if (fieldString.substr(fieldString.size() - 1, 1) == "\r")
	{
		fieldValue = fieldString.substr(valueStartPos, fieldString.size() - 1 - valueStartPos);
	}
	else
	{
		fieldValue = fieldString.substr(valueStartPos);
	}

	return fieldValue;
}

void WebRequest::addParams(const std::string& params)
{
	std::vector<std::string> items;
	StringHelpers::split(params, items, "&");

	std::vector<std::string>::iterator it = items.begin();
	std::vector<std::string>::iterator itEnd = items.end();
	for (; it != itEnd; ++it)
	{
		const std::string& item = *it;

		size_t sep = item.find('=');
		if (sep == std::string::npos)
			continue;

		std::string name = item.substr(0, sep);
		std::string value = item.substr(sep + 1);

		// convert hex encoded strings to native strings

		size_t nHex = 0;
		while ((nHex = value.find('%', nHex)) != std::string::npos)
		{
			std::string strHex = "0x" + value.substr(nHex + 1, 2);
			char cChar = static_cast<char>(strtol(strHex.c_str(), nullptr, 16));
			char szTemp[2];
			memset(szTemp, 0, 2);
			sprintf(szTemp, "%c", cChar);
			std::string strChar(szTemp);

			value.replace(nHex, 3, strChar);
		}

		// + to spaces

		size_t nSpace = 0;
		while ((nSpace = value.find('+', nSpace)) != std::string::npos)
		{
			value.replace(nSpace, 1, " ");
		}

		if (!name.empty() && !value.empty())
		{
			m_aParams[name] = value;
		}
	}
}

void WebRequest::processAuthenticationHeader(const std::string& authorizationString)
{
	size_t sepPos = authorizationString.find(' ');
	if (sepPos == std::string::npos)
	{
		m_headerAuthenticationType = eAuthMalformed;
		return;
	}

	std::string authenticationType = authorizationString.substr(0, sepPos);

	if (authenticationType != "Basic")
	{
		m_headerAuthenticationType = eAuthUnknown;
		return;
	}

	m_headerAuthenticationType = eAuthBasic;

	std::string authorizationToken = authorizationString.substr(sepPos + 1);

	// remove any trailing "\r" char
	if (authorizationToken[authorizationToken.size() - 1] == '\r')
	{
		authorizationToken = authorizationToken.substr(0, authorizationToken.size() - 1);
	}

	authorizationToken = StringHelpers::base64Decode(authorizationToken);

	if (m_headerAuthenticationType == eAuthBasic)
	{
		size_t passSep = authorizationToken.find(':');
		if (passSep == std::string::npos)
		{
			m_headerAuthenticationType = eAuthMalformed;
			return;
		}

		std::string authUser = authorizationToken.substr(0, passSep);
		std::string authPass = authorizationToken.substr(passSep + 1);

		m_authUsername = authUser;
		m_authPassword = authPass;
	}
}

void WebRequest::processCookieHeader(const std::string& cookieString)
{
	std::vector<std::string> items;
	// TODO: this probably isn't completely robust to all types of strings, even if they are hex-encoded...
	StringHelpers::split(cookieString, items, "; ");

	for (const std::string& item : items)
	{
		size_t sep = item.find('=');
		if (sep == std::string::npos)
			continue;

		std::string name = item.substr(0, sep);
		std::string value = item.substr(sep + 1);

		if (!name.empty() && !value.empty())
		{
			m_aCookies[name] = value;
		}
	}
}

File: rs/src/web_request.rs
/*
 WebServe (Rust port)
 Copyright 2021 Peter Pearson.

 Licensed under the Apache License, Version 2.0 (the "License");
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ---------
*/

#![allow(dead_code)]
#![allow(non_snake_case)]

use std::collections::BTreeMap;

use crate::string_helpers;

#[derive(Clone, Copy, Debug)]
#[derive(PartialEq)]
pub enum HTTPVersion {
    HTTPUnknown,
    HTTP09,
    HTTP10,
    HTTP11
}

#[derive(Clone, Copy, Debug)]
#[derive(PartialEq)]
pub enum HTTPRequestType {
    Unknown,
    GET,
    POST,
    HEAD
}

#[derive(Clone, Copy, Debug)]
#[derive(PartialEq)]
pub enum ConnectionType {
    ConnectionUnknown,
    ConnectionClose,
    ConnectionKeepAlive
}

#[derive(Clone, Debug)]
#[derive(PartialEq)]
pub struct WebRequest {
    raw_request:        String,

    // extracted field strings
    user_agent_field:   String,

    // decoded / processed items...
 pub request_type:       HTTPRequestType,
 pub request_version:    HTTPVersion,

 pub path:               String,

 pub host_value:         String,
    connection_value:   String,

 pub connection_type:    ConnectionType,

    params:             BTreeMap<String, String>,
    cookies:            BTreeMap<String, String>,
}

#[derive(Clone, Copy, Debug)]
#[derive(PartialEq)]
pub enum WebRequestParseError {
    InvalidRequest
}

pub type ParseResult = Result<WebRequest, WebRequestParseError>;

impl Default for WebRequest {
    fn default () -> WebRequest {
        WebRequest{raw_request: String::new(), user_agent_field: String::new(),
                    request_type: HTTPRequestType::Unknown, request_version: HTTPVersion::HTTPUnknown,
                    path: String::new(),
                    host_value: String::new(),
                    connection_value: String::new(),
                    connection_type: ConnectionType::ConnectionUnknown,
                    params: BTreeMap::new(), cookies: BTreeMap::new()
                }
    }
}

fn extract_field_item(field_string: &str, start_pos: usize) -> &str {
    let value = &field_string[start_pos..];
    return value;
}

impl WebRequest {

    pub fn create_from_string(request: &str) -> ParseResult {
        let raw_request = request.clone();

        // TODO: doing it this way is very hacky as it means we lose blank lines (end of header, etc)...

        let mut request_lines = raw_request.lines();

        // first line should contain main request of "<COMMAND> <PATH> <HTTP_VER>"...
        let first_line = request_lines.next().unwrap();
        let request_components: Vec<&str> = first_line.splitn(3, ' ').collect();
        if request_components.len() != 3 {
            return Err(WebRequestParseError::InvalidRequest);
        }

        let http_command_string = request_components[0];
        let request_type = match http_command_string {
            "GET" => HTTPRequestType::GET,
            "POST" => HTTPRequestType::POST,
            "HEAD" => HTTPRequestType::HEAD,
            _   => HTTPRequestType::Unknown
        };

        let http_version_string = &request_components[2];
        if http_version_string.len() != 8 {
            return Err(WebRequestParseError::InvalidRequest);
        }
        if !http_version_string.starts_with("HTTP/") {
            return Err(WebRequestParseError::InvalidRequest);
        }
        // advance to actual version part, from 5th char for 3 chars...
        let http_version_string_token = &http_version_string[5..][..3];

        let request_version = match http_version_string_token {
            "0.9" => HTTPVersion::HTTP09,
            "1.0" => HTTPVersion::HTTP10,
            "1.1" => HTTPVersion::HTTP11,
            _     => HTTPVersion::HTTPUnknown
        };

        let mut connection_type = ConnectionType::ConnectionUnknown;

        if request_version == HTTPVersion::HTTP11 {
            connection_type = ConnectionType::ConnectionKeepAlive;
        }

        let mut path = request_components[1].to_string();
        // TODO: look for GET params after '?'

        let mut params = BTreeMap::new();

        let question_mark_pos = path.find('?');
        if question_mark_pos.is_some() {
            if request_type == HTTPRequestType::GET {
                let get_params_string = &path[question_mark_pos.unwrap()+1..].to_string();
                params = WebRequest::process_get_params(&get_params_string);
            }
            path = path[..question_mark_pos.unwrap()].to_string();
        }

        let mut found_host = false;
        let mut found_connection = false;
        let mut found_cookie = false;

        let mut host_value = String::new();
        let mut cookies = BTreeMap::new();

        // now process all other lines...
        // TODO: something more efficient than this...
        // TODO: and might need to be case-insensitive?
        for line in request_lines {
            if !found_host && line.starts_with("Host:") {
                host_value = extract_field_item(&line, 5 + 1).to_string();
                found_host = true;
            }
            else if !found_cookie && line.starts_with("Cookie:") {
                let cookie_value = extract_field_item(&line, 7 + 1).to_string();
                cookies = WebRequest::process_cookie_string(&cookie_value);
                found_cookie = true;
            }
            else if !found_connection && line.starts_with("Connection:") {
                let connection_value = extract_field_item(&line, 11 + 1).to_string();
                match connection_value.as_str() {
                    "close" => {
                        connection_type = ConnectionType::ConnectionClose;
                    }
                    "keep-alive" => {
                        connection_type = ConnectionType::ConnectionKeepAlive;
                    }
                    _ => {}
                }
                found_connection = true;
            }
        }

        Ok(WebRequest { raw_request: raw_request.to_string(), request_type, request_version, path, host_value, connection_type, params, cookies, ..Default::default() })
    }

    pub fn has_param(&self, param: &str) -> bool {
        return self.params.contains_key(param);
    }

    pub fn get_param(&self, param: &str) -> String {
        let mut ret_value = String::new();

        if self.params.contains_key(param) {
            ret_value = self.params.get(param).unwrap().clone();
        }

        return ret_value;
    }

    pub fn get_param_as_uint(&self, param: &str, default: u32) -> u32 {
        let result = self.params.get(param);
        match result {
            None => return default,
            Some(val) => {
                let uint_value = val.parse::<u32>().unwrap();
                return uint_value;
            }
        }
    }

    pub fn get_params_as_GET_string(&self, ignore_pagination_params: bool) -> String {
        let mut params_string = String::new();

        // TODO: can we use iterators + fold/map for this?
        for (k,v) in &self.params {
            if ignore_pagination_params && (k == "perPage" || k == "startIndex") {
                continue;
            }

            if !params_string.is_empty() {
                params_string.push('&');
            }

            // TODO: hex/string/encoding of things...
            params_string.push_str(&format!("{}={}", k, v));
        }

        return params_string;
    }

    pub fn has_cookie(&self, cookie: &str) -> bool {
        return self.cookies.contains_key(cookie);
    }

    pub fn get_cookie(&self, cookie: &str) -> String {
        let mut ret_value = String::new();

        if self.cookies.contains_key(cookie) {
            ret_value = self.cookies.get(cookie).unwrap().clone();
        }

        return ret_value;
    }

    pub fn get_cookie_as_uint(&self, cookie: &str, default: u32) -> u32 {
        let result = self.cookies.get(cookie);
        match result {
            None => return default,
            Some(val) => {
                let uint_value = val.parse::<u32>().unwrap();
                return uint_value;
            }
        }
    }

    pub fn get_param_or_cookie_as_uint(&self, param_name: &str, cookie_name: &str, default: u32) -> u32 {
        let result = self.params.get(param_name);
        match result {
            Some(val) => {
                let uint_value = val.parse::<u32>().unwrap();
                return uint_value;
            }
            None => {
                return self.get_cookie_as_uint(cookie_name, default);
            }   
        }
    }

    fn process_get_params(params_string: &str) -> BTreeMap<String, String> {
        let mut params = BTreeMap::new();

        for item in params_string.split('&') {
            if let Some((key, val)) = item.split_once('=') {
                
                let val = &string_helpers::simple_decode_string(&val);

                if !key.is_empty() && !val.is_empty() {
                    params.insert(key.to_string(), val.to_string());
                }
            }
        }

        return params;
    }

    fn process_cookie_string(cookie_string: &str) -> BTreeMap<String, String> {
        let mut cookies = BTreeMap::new();

        // TODO: this probably isn't completely robust to all types of strings, even if they are hex-encoded...
        for item in cookie_string.split("; ") {
            if let Some((key, val)) = item.split_once('=') {
                if !key.is_empty() && !val.is_empty() {
                    cookies.insert(key.to_string(), val.to_string());
                }
            }
        }

        return cookies;
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_main_invalid() {
        let req_string = "/ g HTTP1.1\r\n".to_string();

        let web_request = WebRequest::create_from_string(&req_string);

        assert!(web_request.is_err(), "web_request should be an error...");
        assert_eq!(web_request, Err(WebRequestParseError::InvalidRequest));
    }

    #[test]
    fn request_main_version1() {
        let req_string = "GET / HTTP/1.1\r\nHost: test.com\r\n".to_string();

        let web_request = WebRequest::create_from_string(&req_string).unwrap();

        assert_eq!(web_request.request_type, HTTPRequestType::GET);
        assert_eq!(web_request.path, "/");
        assert_eq!(web_request.request_version, HTTPVersion::HTTP11);

        assert_eq!(web_request.host_value, "test.com");
    }

    #[test]
    fn request_main_version2() {
        let req_string = "POST /submit?force=true HTTP/1.1\r\nHost: test.com\r\n";

        let web_request = WebRequest::create_from_string(&req_string).unwrap();

        assert_eq!(web_request.request_type, HTTPRequestType::POST);
        assert_eq!(web_request.path, "/submit");
        assert_eq!(web_request.request_version, HTTPVersion::HTTP11);

        assert_eq!(web_request.host_value, "test.com");
    }

    

}
